pandas/plotting/_matplotlib/groupby.py

```diff
@@ -0,0 +1,136 @@
+from typing import (
+    Dict,
+    Optional,
+    Union,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    FrameOrSeriesUnion,
+    IndexLabel,
+)
+
+from pandas.core.dtypes.missing import remove_na_arraylike
+
+from pandas import (
+    DataFrame,
+    MultiIndex,
+    Series,
+    concat,
+)
+
+
+def create_iter_data_given_by(
+    data: DataFrame, by: Optional[IndexLabel] = None, kind: str = "hist"
+) -> Union[DataFrame, Dict[str, FrameOrSeriesUnion]]:
+    """
+    Create data for iteration given `by` is assigned or not, and it is only
+    used in both hist and boxplot.
+
+    If `by` is assigned, return a dictionary of DataFrames in which the key of
+    dictionary is the values in groups.
+    If `by` is not assigned, return input as is, and this preserves current
+    status of iter_data.
+
+    Parameters
+    ----------
+    data: reformatted grouped data from `_compute_plot_data` method.
+    by: list or None, value assigned to `by`.
+    kind: str, plot kind. This function is only used for `hist` and `box` plots.
+
+    Returns
+    -------
+    iter_data: DataFrame or Dictionary of DataFrames
+
+    Examples
+    --------
+    If `by` is assigned:
+
+    >>> import numpy as np
+    >>> tuples = [('h1', 'a'), ('h1', 'b'), ('h2', 'a'), ('h2', 'b')]
+    >>> mi = MultiIndex.from_tuples(tuples)
+    >>> value = [[1, 3, np.nan, np.nan],
+    ...          [3, 4, np.nan, np.nan], [np.nan, np.nan, 5, 6]]
+    >>> data = DataFrame(value, columns=mi)
+    >>> create_iter_data_given_by(data, by=["col1"])
+    {'h1': DataFrame({'a': [1, 3, np.nan], 'b': [3, 4, np.nan]}),
+     'h2': DataFrame({'a': [np.nan, np.nan, 5], 'b': [np.nan, np.nan, 6]})}
+    """
+    if kind == "hist":
+        level = 0
+    elif kind == "box":
+        level = 1
+    else:
+        raise ValueError("This function is only used for hist and box plot")
```

Comment:
Sure, updated accordingly! 

This `else` branch won't be reached at all, because this function is only used `hist` and `boxplot` plot implementation and not being called anywhere else.